### ISSUE TITLE
Add results below the line level

### ISSUE THREAD

**Issue by bertsky** [OCR-D/ocrd_calamari#9]
Other OCR processors like [Tesseract](OCR-D/ocrd_tesserocr) and [Ocropy](cisocrgroup/ocrd_cis) go out of their way to provide annotation on the `Word` or even `Glyph` level if requested. (They use the line result with precise character-wise confidences and relative positions for that.)

I know this will be a bit harder for Calamari with its text post-processing and voting capabilities. But IMO this is worth the effort.

**Comment by mikegerber** [comment]
I'll have to investigate Calamari's extended prediction data for this. It's not super high on my priority list though, because I currently work exclusively with line segmentation and line OCR.

**Comment by mikegerber** [comment]
I met @cneud and he also told me that word-level annotation is needed. I have to think this through and review how's it's done with Tesseract etc.

**Comment by cneud** [comment]
Yes, for productive use in digital libraries, word level segmentation is a MUST - the word bounding box information is used for highlighting keyword search results in a viewer.

**Comment by mikegerber** [comment]
master now contains early support for this. It's useful for alternative predictions (#28) and for per-glyph confidence values but not if you need proper coordinates.

**Comment by mikegerber** [comment]
@bertsky I'm going to close this issue now, while I might/will improve it in the future in these ways:

* crop segmentation to the line segmentation. (I currently use the height of the line image as the height for word+glyph segmentation)
* adjust word segmentation to the expectations of the OCR-D validators

I'm probably not going to improve the segmentation coordinates further because that would mean introducing a significantly complex algorithm to a processor that should be restricted to provide an interface to the Calamari engine.

In other words I see these use cases solved:
* Provide word segmentation for highlighting
* Provide glyph segmentation (with confidences and alternatives) for post-correction
And these are out of scope:
* No good coordinates for further image based processing (i.e. cropping out characters)

**Comment by bertsky** [comment]
Excellent – thanks!

> I might/will improve it in the future in these ways:
> 
>     * crop segmentation to the line segmentation. (I currently use the height of the line image as the height for word+glyph segmentation)

Both our Tesseract/ocrd_tesserocr and Ocropus/ocrd_cis wrapper use the vertical bounds of the line image, too.

> 
>     * adjust word segmentation to the expectations of the OCR-D validators

Should not be too difficult – just use `sentence.split()` instead of `uniseg.wordbreak.words(sentence)`.

> And these are out of scope:
> 
>     * No good coordinates for further image based processing (i.e. cropping out glyphs)

I absolutely agree.

**Comment by bertsky** [comment]
@mikegerber I think it would make sense to offer a `textequiv_level` parameter to control this (line, word, glyph) as in `ocrd-tesserocr-recognize` and `ocrd-cis-ocropy-recognize`. Do you want me to open a new issue for that?

Also, what is the current probabiliity cut-off criterion? Maybe this could be parameterized as well (in relative terms, as a fraction over best, and/or in absolute terms, as a maximum number of alternatives)...

**Comment by mikegerber** [comment]
> @mikegerber I think it would make sense to offer a `textequiv_level` parameter to control this (line, word, glyph) as in `ocrd-tesserocr-recognize` and `ocrd-cis-ocropy-recognize`. Do you want me to open a new issue for that?

Please do. We should discuss if this is to control "level of operation" or for "granularity of output" there.
 
> Also, what is the current probabiliity cut-off criterion? Maybe this could be parameterized as well (in relative terms, as a fraction over best, and/or in absolute terms, as a maximum number of alternatives)...

There is no cut-off criterion yet, I just sort all results. There is probably a very small one in Calamari itself. The smallest I've seen is in the range of around 0.5e-05.

**Comment by bertsky** [comment]
> Please do. We should discuss if this is to control "level of operation" or for "granularity of output" there.

See #31 

> There is no cut-off criterion yet, I just sort all results. There is probably a very small one in Calamari itself. The smallest I've seen is in the range of around 0.5e-05.

Fair enough. As long as it does not bloat arbitrarily. And 5e-6 should be well below minimum a-priori probability for reasonably distributed character sets. (I wonder how Chinese characters would be modelled though; Tesseract splits up into a sequence code points to reduce the sparsity and computational cost of the output layer...)
